Source/WebCore/platform/CaretAnimation.h

```diff
@@ -0,0 +1,129 @@
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+
+#include "Document.h"
+#include "ReducedResolutionSeconds.h"
+#include "Timer.h"
+#include <wtf/FastMalloc.h>
+
+namespace WebCore {
+
+class Page;
+class CaretAnimation;
+
+class CaretAnimationClient {
+public:
+    virtual ~CaretAnimationClient() = default;
+
+    virtual void caretAnimationDidUpdate(CaretAnimation&) { }
+    virtual void caretAnimationWillStart(CaretAnimation&) { }
+    virtual void caretAnimationDidEnd(CaretAnimation&) { }
+};
+
+class CaretAnimation {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    enum class Type {
+        Simple
+    };
+
+    CaretAnimation(Type animationType, CaretAnimationClient& client, Document* document) 
+        : m_animationType(animationType)
+        , m_caretBlinkTimer(*this, &CaretAnimation::scheduleAnimation)
+        , m_client(client)
+        , m_document(document) 
+    { }
+
+    virtual ~CaretAnimation() = default;
+
+    Type type() const { return m_animationType; }
+
+    virtual void start(ReducedResolutionSeconds currentTime) = 0;
+
+    virtual void stop()
+    {
+        if (!m_isActive)
+            return;
+        didEnd();
+    }
+
+    virtual bool isActive() const { return m_isActive; }
+
+    virtual void serviceAnimation(ReducedResolutionSeconds) = 0;
+
+    void serviceCaretAnimation(ReducedResolutionSeconds);
+
+    void scheduleAnimation();
+
+    virtual String debugDescription() const = 0;
+
+    void setCaretBlinkingSuspended(bool suspended) { m_isCaretBlinkingSuspended = suspended; }
+    bool isCaretBlinkingSuspended() const { return m_isCaretBlinkingSuspended; }
+
+    virtual bool visible() const = 0;
+    virtual void setVisible(bool) = 0;
+
+protected:
+    void didStart(ReducedResolutionSeconds currentTime, Seconds interval)
+    {
+        m_startTime = currentTime;
+        m_isActive = true;
+        m_client.caretAnimationWillStart(*this);
+        m_caretBlinkTimer.startRepeating(interval);
+    }
+
+    void didEnd()
+    {
+        m_isActive = false;
+        m_client.caretAnimationDidEnd(*this);
+        m_caretBlinkTimer.stop();
+    }
+
+    Seconds timeSinceStart(ReducedResolutionSeconds currentTime) const
+    {
+        return currentTime - m_startTime;
+    }
+
+    Page* page() const;
+
+    const Type m_animationType;
+    bool m_isActive { false };
```

Comment:
Re-order these members to optimize packing (put the bools next to each other for example).